### PR TITLE
Stop BCELoss from returning negative results

### LINKED ISSUE #2866: CrossEntropyLoss is negative
what’s the formula of CrossEntropyLoss used in pytorch?
I used CrossEntropyLoss, but the loss value is negative. generally speaking, it should be positive.
so what’s wrong?

### PR DESCRIPTION
**Pull request by li-roy**
Fix #2866.

cc: @zou3519

### CODE REVIEW

**Comment by fmassa** [comment]
I'm not sure this is the right fix.
What happens when x=1 and  y=1?

**Comment by zou3519** [comment]
As I understand it, the formula as it is right now is `y log (x + eps) + (1-y) log (1 - x + eps)`
Maybe it's better to don't use the `eps` and instead check if the input to log is 0?

In the original issue it looks like the error is `1.00000e-12 * -1.0001` which is really not too bad unless the sign of the output really matters.

**Comment by ezyang** [comment]
Needs test.

File: test/test_nn.py [hunk 1]
```diff
@@ -4116,6 +4116,15 @@ def func(root):
         gradcheck(func, [v])
         gradgradcheck(func, [v])
 
+    def test_bce_loss_always_nonnegative(self):
+        target = torch.ones(5)
+        input = torch.ones(5)
+        assert ((nn.BCELoss()(input, target) < 0).sum() == 0)
```
Comment (zou3519) [inline test/test_nn.py]: nit: use self.assertEqual maybe so we get a nice error message?

File: aten/src/THNN/generic/BCECriterion.c [hunk 2]
```diff
@@ -4,6 +4,13 @@
 
 #define EPS 1e-12
 
+static inline real safe_log(real a) {
+  if (a == 0.) {
+    return log(EPS);
+  }
+  return log(a);
```
[the remark this answers is not in the document]
Comment (li-roy) [reply aten/src/THNN/generic/BCECriterion.c]: I think log is just a macro that calls the correct function for each type.

http://en.cppreference.com/w/c/numeric/math/log

**Review by zou3519 — APPROVED** [review]
lgtm

**Review by soumith — APPROVED** [review]
